src/Region.h

```diff
@@ -0,0 +1,172 @@
+
+public:
+  virtual ~Region();
+
+  /* Functions for constructing the Region / other Regions */
+  virtual void addNode(Region* node, bool clone=true);
+  void removeHalfspace(Surface* surface, int halfspace);
+  regionType getRegionType();
+  void setParentRegion(Region* node);
+  Region* getParentRegion();
+
+  /* Getter functions */
+  virtual std::vector<Region*> getNodes();
+  virtual std::vector<Region*> getAllNodes();
+  virtual std::map<int, Halfspace*> getAllSurfaces();
+
+  /* Worker functions */
+  virtual double getMinX();
+  virtual double getMaxX();
+  virtual double getMinY();
+  virtual double getMaxY();
+  virtual double getMinZ();
+  virtual double getMaxZ();
+  virtual boundaryType getMinXBoundaryType();
+  virtual boundaryType getMaxXBoundaryType();
+  virtual boundaryType getMinYBoundaryType();
+  virtual boundaryType getMaxYBoundaryType();
+  virtual boundaryType getMinZBoundaryType();
+  virtual boundaryType getMaxZBoundaryType();
+  virtual bool containsPoint(Point* point) =0;
+  virtual double minSurfaceDist(Point* point, double azim, double polar=M_PI_2);
+  virtual double minSurfaceDist(LocalCoords* coords);
+  virtual Region* clone();
+};
+
+/**
+ * @class Intersection Region.h "src/Region.h"
+ * @brief An intersection of two or more Regions.
+ */
+class Intersection : public Region {
+
+public:
+  Intersection();
+  bool containsPoint(Point* point);
+};
+
+/**
+ * @class Union Region.h "src/Region.h"
+ * @brief A union of two or more Regions.
+ */
+class Union : public Region {
+
+public:
+  Union();
+  bool containsPoint(Point* point);
+};
+
+/**
+ * @class Complement Region.h "src/Region.h"
+ * @brief A complement of a Region.
+ */
+class Complement : public Region {
+
+public:
+  Complement();
+  bool containsPoint(Point* point);  
+};
+
+/**
+ * @class Halfspace Region.h "src/Region.h"
+ * @brief A positive or negative halfspace Region.
+ */
+class Halfspace : public Region {
+  
+public:
+
+  /** A pointer to the Surface object */
+  Surface* _surface;
+
+  /** The halfspace associated with this surface */
+  int _halfspace;
+
+  Halfspace(int halfspace, Surface* surface);
+  Halfspace* clone();
+  Surface* getSurface();
+  int getHalfspace();
+  void reverseHalfspace();
+  std::map<int, Halfspace*> getAllSurfaces();
+  double getMinX();
+  double getMaxX();
+  double getMinY();
+  double getMaxY();
+  double getMinZ();
+  double getMaxZ();
+  boundaryType getMinXBoundaryType();
+  boundaryType getMaxXBoundaryType();
+  boundaryType getMinYBoundaryType();
+  boundaryType getMaxYBoundaryType();
+  boundaryType getMinZBoundaryType();
+  boundaryType getMaxZBoundaryType();
+  bool containsPoint(Point* point);
+  double minSurfaceDist(Point* point, double azim, double polar=M_PI_2);
+  double minSurfaceDist(LocalCoords* coords);
+};
+
+/**
+ * @class RectangularPrism Region.h "src/Region.h"
+ * @brief An infinite rectangular prism aligned with the z-axis.
+ */
+class RectangularPrism : public Intersection {
```

Comment:
It will be better to make it compatible for 3D cube region like RectangularPrism(double width_x, double width_y,
       double origin_x=0., double origin_y=0., double width_y=inf, double origin_z=0.);